src/rust/engine/src/watch.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::collections::HashSet;
+use std::path::PathBuf;
+use std::sync::{Arc, Weak};
+use std::thread;
+use std::time::Duration;
+
+use boxfuture::{BoxFuture, Boxable};
+use crossbeam_channel::{self, Receiver, RecvTimeoutError, TryRecvError};
+use futures01 as futures;
+use futures01::future::Future;
+use log::{debug, error, info, warn};
+use notify::{RecommendedWatcher, RecursiveMode, Watcher};
+use parking_lot::Mutex;
+
+use graph::{Graph, InvalidationResult};
+use logging;
+
+use crate::nodes::NodeKey;
+
+///
+/// An InvalidationWatcher maintains a Thread that receives events from a notify Watcher.
+///
+/// If the spawned Thread exits for any reason, InvalidationWatcher::running() will return False,
+/// and the caller should create a new InvalidationWatcher (or shut down, in some cases). Generally
+/// this will mean polling.
+///
+/// TODO: Need the above polling
+///
```

Comment:
Thanks a lot for getting the tests in place here.

It seems like it might be possible to remove the references to `Graph` (and thus `NodeKey`) here if `InvalidationWatcher` exposed a small trait with one method: `invalidate_from_roots`. Then your test would use a custom implementation of that trait.

Since you already have this working, maybe you could leave a TODO here about moving `Graph` into a generic parameter?

Reply:
Can you re-impl a trait if it has already been impl'd in antoher module? What I'm imagining doing here is declare a trait in watch.rs which has a generic graph parameter that it takes in the `invalidate_from_roots` method. impl that Trait for the the InvalidationWatcher where the parameter is a Graph. Then in tests impl the Trait where the parameter is something else, probably a channel which I use to report invalidation events to the test.

I think I'll just leave this as a TODO, because it seems somewhat complicated.

Reply:
That sounds more complex than I was thinking. I was suggesting roughly:
```
trait Invalidatable {
  fn invalidate_from_roots(&mut self) -> ???;
}

impl InvalidationWatcher {
  pub fn new<I: Invalidatable>(
    graph: Weak<I>,
    build_root: PathBuf,
  ) -> Result<InvalidationWatcher, String> {
```

Reply:
TIL. Thanks